File: src/SchoolLunchMenu/Services/DayLabelFetchService.cs
namespace SchoolLunchMenu.Services;

using System.Net.Http;
using System.Text.RegularExpressions;
using Microsoft.Extensions.Logging;

/// <summary>
/// Scrapes day label entries (Red Day / White Day) from the ISD194 Finalsite CMS calendar page.
/// </summary>
public partial class DayLabelFetchService : IDayLabelFetchService
{
    private const string CalendarUrl = "https://cms.isd194.org/news-and-events/calendar";

    private readonly HttpClient _httpClient;
    private readonly ILogger<DayLabelFetchService> _logger;

    public DayLabelFetchService(HttpClient httpClient, ILogger<DayLabelFetchService> logger)
    {
        _httpClient = httpClient;
        _logger = logger;
    }

    /// <inheritdoc />
    public async Task<DayLabelFetchResult> FetchAsync()
    {
        _logger.LogInformation("Fetching day labels from {Url}", CalendarUrl);

        var html = await _httpClient.GetStringAsync(CalendarUrl);

        var entries = ParseDayLabels(html);

        var distinctLabels = entries
            .Select(e => e.Label)
            .Distinct(StringComparer.OrdinalIgnoreCase)
            .ToList();

        _logger.LogInformation("Found {Count} day label entries with {Distinct} distinct labels",
            entries.Count, distinctLabels.Count);

        return new DayLabelFetchResult
        {
            Entries = entries,
            DistinctLabels = distinctLabels
        };
    }

    /// <summary>
    /// Parses day label entries from the Finalsite CMS calendar HTML.
    /// Each calendar day is a <c>fsCalendarDate</c> div with data-day, data-year, data-month attributes.
    /// Events within are <c>fsCalendarEventTitle</c> anchors with a title attribute.
    /// Months are 0-indexed in the HTML (January = 0).
    /// </summary>
    private List<(DateOnly Date, string Label)> ParseDayLabels(string html)
    {
        var results = new List<(DateOnly Date, string Label)>();

        // Split HTML into date blocks by finding each fsCalendarDate section.
        // We track the current date context and look for event titles within each block.
        var dateMatches = DateBlockRegex().Matches(html);

        foreach (Match dateMatch in dateMatches)
        {
            if (!int.TryParse(dateMatch.Groups["day"].Value, out var day) ||
                !int.TryParse(dateMatch.Groups["year"].Value, out var year) ||
                !int.TryParse(dateMatch.Groups["month"].Value, out var month0))
                continue;

            // Finalsite months are 0-indexed
            var month = month0 + 1;

            if (month is < 1 or > 12 || day < 1 || day > DateTime.DaysInMonth(year, month))
                continue;

            var date = new DateOnly(year, month, day);

            // Find the end of this day block (next fsCalendarDate or end of string)
            var blockStart = dateMatch.Index;
            var nextDateMatch = dateMatches.Cast<Match>()
                .FirstOrDefault(m => m.Index > blockStart);
            var blockEnd = nextDateMatch?.Index ?? html.Length;
            var block = html[blockStart..blockEnd];

            // Find event titles within this block
            var titleMatches = EventTitleRegex().Matches(block);
            foreach (Match titleMatch in titleMatches)
            {
                var title = titleMatch.Groups["title"].Value.Trim();

                // Match "Red Day", "White Day", or similar rotation labels
                if (DayLabelPattern().IsMatch(title))
                {
                    results.Add((date, title));
                }
            }
        }

        results.Sort((a, b) => a.Date.CompareTo(b.Date));
        return results;
    }

    // Matches fsCalendarDate div with data attributes (handles any attribute order)
    [GeneratedRegex(
        @"fsCalendarDate[^>]*?data-day=""(?<day>\d+)""[^>]*?data-year=""(?<year>\d+)""[^>]*?data-month=""(?<month>\d+)""",
        RegexOptions.Singleline)]
    private static partial Regex DateBlockRegex();

    // Matches fsCalendarEventTitle anchor with title attribute
    [GeneratedRegex(
        @"fsCalendarEventTitle[^>]*?title=""(?<title>[^""]+)""",
        RegexOptions.Singleline)]
    private static partial Regex EventTitleRegex();

    // Matches day label patterns like "Red Day", "White Day"
    [GeneratedRegex(@"^(?:Red|White|Blue|Gold|Green|Silver|Black|Orange|Purple|Day\s*[A-Z]|[A-Z])\s*Day$",
        RegexOptions.IgnoreCase)]
    private static partial Regex DayLabelPattern();
}

File: school-lunch-menu.Mac/SchoolLunchMenu/Services/DayLabelFetchService.swift
import Foundation
import os

/// Scrapes day label entries (Red Day / White Day) from the ISD 194 Finalsite CMS calendar page.
struct DayLabelFetchService {
    private static let calendarUrl = "https://cms.isd194.org/news-and-events/calendar"
    private static let logger = Logger(subsystem: "com.schoollunchmenu", category: "DayLabelFetchService")

    /// Result of a day label fetch operation.
    struct FetchResult {
        /// Date-label pairs found on the calendar page, sorted by date.
        let entries: [(date: DateComponents, label: String)]
        /// Distinct label names found, in order of first appearance.
        let distinctLabels: [String]
    }

    /// Fetches day labels from the ISD 194 CMS calendar page.
    func fetch() async throws -> FetchResult {
        Self.logger.info("Fetching day labels from \(Self.calendarUrl)")

        guard let url = URL(string: Self.calendarUrl) else {
            throw URLError(.badURL)
        }

        let (data, _) = try await URLSession.shared.data(from: url)
        guard let html = String(data: data, encoding: .utf8) else {
            throw URLError(.cannotDecodeContentData)
        }

        let entries = parseDayLabels(html: html)

        // Collect distinct labels in order of first appearance
        var seen = Set<String>()
        var distinctLabels: [String] = []
        for entry in entries {
            let lower = entry.label.lowercased()
            if !seen.contains(lower) {
                seen.insert(lower)
                distinctLabels.append(entry.label)
            }
        }

        Self.logger.info("Found \(entries.count) day label entries with \(distinctLabels.count) distinct labels")

        return FetchResult(entries: entries, distinctLabels: distinctLabels)
    }

    /// Parses day label entries from Finalsite CMS calendar HTML.
    ///
    /// Each calendar day is a `fsCalendarDate` div with `data-day`, `data-year`, `data-month` attributes.
    /// Events within are `fsCalendarEventTitle` anchors with a `title` attribute.
    /// Months are 0-indexed in the HTML (January = 0).
    private func parseDayLabels(html: String) -> [(date: DateComponents, label: String)] {
        var results: [(date: DateComponents, label: String)] = []

        // Match fsCalendarDate divs with data attributes
        let datePattern = #"fsCalendarDate[^>]*?data-day="(\d+)"[^>]*?data-year="(\d+)"[^>]*?data-month="(\d+)""#
        guard let dateRegex = try? NSRegularExpression(pattern: datePattern, options: .dotMatchesLineSeparators) else {
            return results
        }

        let nsHtml = html as NSString
        let dateMatches = dateRegex.matches(in: html, range: NSRange(location: 0, length: nsHtml.length))

        // Match event titles
        let titlePattern = #"fsCalendarEventTitle[^>]*?title="([^"]+)""#
        guard let titleRegex = try? NSRegularExpression(pattern: titlePattern, options: .dotMatchesLineSeparators) else {
            return results
        }

        // Day label pattern: "Red Day", "White Day", etc.
        let labelPattern = #"^(?:Red|White|Blue|Gold|Green|Silver|Black|Orange|Purple|Day\s*[A-Z]|[A-Z])\s*Day$"#
        guard let labelRegex = try? NSRegularExpression(pattern: labelPattern, options: .caseInsensitive) else {
            return results
        }

        for (i, dateMatch) in dateMatches.enumerated() {
            guard dateMatch.numberOfRanges >= 4,
                  let dayRange = Range(dateMatch.range(at: 1), in: html),
                  let yearRange = Range(dateMatch.range(at: 2), in: html),
                  let month0Range = Range(dateMatch.range(at: 3), in: html),
                  let day = Int(html[dayRange]),
                  let year = Int(html[yearRange]),
                  let month0 = Int(html[month0Range]) else {
                continue
            }

            // Finalsite months are 0-indexed
            let month = month0 + 1

            guard month >= 1, month <= 12, day >= 1 else { continue }

            // Find the block of HTML for this date (until the next date or end of string)
            let blockStart = dateMatch.range.location
            let blockEnd: Int
            if i + 1 < dateMatches.count {
                blockEnd = dateMatches[i + 1].range.location
            } else {
                blockEnd = nsHtml.length
            }

            let blockRange = NSRange(location: blockStart, length: blockEnd - blockStart)
            let block = nsHtml.substring(with: blockRange)

            // Find event titles within this block
            let titleMatches = titleRegex.matches(in: block, range: NSRange(location: 0, length: (block as NSString).length))
            for titleMatch in titleMatches {
                guard titleMatch.numberOfRanges >= 2,
                      let titleRange = Range(titleMatch.range(at: 1), in: block) else {
                    continue
                }

                let title = String(block[titleRange]).trimmingCharacters(in: .whitespaces)

                // Check if this matches a day label pattern
                let titleNSRange = NSRange(location: 0, length: (title as NSString).length)
                if labelRegex.firstMatch(in: title, range: titleNSRange) != nil {
                    let components = DateComponents(year: year, month: month, day: day)
                    results.append((date: components, label: title))
                }
            }
        }

        // Sort by date
        results.sort { a, b in
            let aDate = (a.date.year ?? 0) * 10000 + (a.date.month ?? 0) * 100 + (a.date.day ?? 0)
            let bDate = (b.date.year ?? 0) * 10000 + (b.date.month ?? 0) * 100 + (b.date.day ?? 0)
            return aDate < bDate
        }

        return results
    }
}
